[PRIV-128] Aggregate responses

cedric-cordenier opened the pull request:
* move MethodX and SignedOCRResponse constants to vault/capability.go
* introduce an aggregator abstraction into the vault gateway handler. This aggregates based on a) signatures first, and b) quorum second. We fallback to mode because some requests could be returned from the capability DON without going through consensus and won't have signatures. In case we do have signatures and we fallback anyway, we exclude the signatures from the digest since not all responses are guaranteed to have the same signatures.

cedric-cordenier committed:
WIP

cedric-cordenier committed:
WIP

cedric-cordenier committed:
WIP

cedric-cordenier committed:
Merge branch 'develop' into PRIV-128-aggregate-responses

cedric-cordenier committed:
WIP

cedric-cordenier committed:
Merge branch 'develop' into PRIV-128-aggregate-responses

cedric-cordenier committed:
WIP

cedric-cordenier committed:
WIP

cedric-cordenier committed:
Merge branch 'develop' into PRIV-128-aggregate-responses

jmank88 commented on core/services/gateway/handlers/vault/handler_test.go:
```diff
@@ -45,14 +48,38 @@ func setupHandler(t *testing.T) (handlers.Handler, chan handlers.UserCallbackPay
 	require.NoError(t, err)
 
-	handler, err := NewHandler(methodConfig, donConfig, don, lggr)
+	// TODO
+	handler, err := NewHandler(methodConfig, donConfig, don, nil, lggr)
+	handler.aggregator = &mockAggregator{}
 	require.NoError(t, err)
```
More to do here?
```suggestion
	handler, err := NewHandler(methodConfig, donConfig, don, nil, lggr)
	require.NoError(t, err)
	handler.aggregator = &mockAggregator{}
```

cedric-cordenier replied:
Thanks 🙏 Actually this is sorted now, I've removed the TODO

jmank88 commented on core/services/registrysyncer/local_registry.go:
```diff
@@ -171,6 +171,62 @@ func (l *LocalRegistry) NodeByPeerID(ctx context.Context, peerID types.PeerID) (
+			}
+
+			nodes = append(nodes, node)
+
+			fmt.Printf("d: %+v\n", d)
+		}
+		(&d).Nodes = nodes
+	}
+
+	fmt.Printf("foundDONs: %+v\n", foundDONs)
```
```suggestion
		}
		(&d).Nodes = nodes
	}

```

cedric-cordenier replied:
Ugh, thanks for catching!

cedric-cordenier committed:
Remove trace logs + TODO

cedric-cordenier committed:
Remove trace logs + TODO

krehermann commented on core/capabilities/vault/capability.go:
```diff
@@ -131,11 +130,33 @@ func (s *Capability) handleRequest(ctx context.Context, requestID string, reques
+	uniqueIDs := map[string]bool{}
+	for _, req := range request.EncryptedSecrets {
+		if req.Id == nil {
+			return nil, errors.New("secret ID must not be nil")
```
if i where the caller, i would want to know the index of the error, assuming that not a security concern of some sort

cedric-cordenier replied:
Good point, will add :)

krehermann commented on core/capabilities/vault/models.go:
```diff
@@ -2,18 +2,49 @@ package vault
 
+const (
+	// Note: any addition to this list should be reflected in
+	// HandlerTypeForMethod in handler_factory.go
```
what the reasoning for copying the values rather than importing shared defn?

cedric-cordenier replied:
Good point -- I've refactored so there's a variable with the list of methods as well as individual method consts

krehermann commented on core/capabilities/vault/models.go:
```diff
@@ -2,18 +2,49 @@ package vault
+type SignedOCRResponse struct {
+	Error      string          `json:"error"`
+	Payload    json.RawMessage `json:"payload"`
+	Context    []byte          `json:"__context"`
```
what's up with the __ in the json names?

cedric-cordenier replied:
I was being smart I think -- the original idea was that these were internal fields to be used to validate a signature, but I've removed it now

krehermann commented on core/capabilities/vault/models.go:
```diff
@@ -2,18 +2,49 @@ package vault
+}
+
+func (r *SignedOCRResponse) String() string {
+	return fmt.Sprintf("SignedOCRResponse { Error: %s, Payload: %s, Context: <[%d]byte blob>, Signatures: <[%d][]byte blob>}", r.Error, string(r.Payload), len(r.Context), len(r.Signatures))
```
this is safe for logs? the ctx doesn't have any sensitive data?

bolekk replied:
context doesn't but payload does, right?

cedric-cordenier replied:
This was mainly being used for debugging (and is copied from the existing implementation; this PR just moves it). We don't currently print out the SignedOCRResponse anywhere.

prashantkumar1982 replied:
I don't think this would have any sensitive data.
All responses are just dealing with listing or reading secrets ids, not the secret themselves.

krehermann commented on core/capabilities/vault/models.go:
```diff
@@ -100,3 +131,58 @@ func (r *Response) RequestID() string {
+		return fmt.Errorf("not enough signatures: expected min %d, got %d", minRequired, len(resp.Signatures))
+	}
+
+	// The context contains:
```
nice, thank you

krehermann replied:
if ctx really ought to be >=64; please add comment about what is (optionally) contained in [64:]

krehermann commented on core/capabilities/vault/models.go:
```diff
@@ -100,3 +131,58 @@ func (r *Response) RequestID() string {
 }
+
+func ValidateSignatures(resp *SignedOCRResponse, allowedSigners []common.Address, minRequired int) error {
+	if len(resp.Context) < 64 {
```
is it really less than? or should it be equal 64? the parse below is only looking upto 64

krehermann commented on core/scripts/gateway/run_gateway.go:
```diff
@@ -48,7 +48,7 @@ func main() {
 	lggr, _ := logger.NewLogger()
 
-	handlerFactory := gateway.NewHandlerFactory(nil, nil, nil, lggr)
+	handlerFactory := gateway.NewHandlerFactory(nil, nil, nil, nil, lggr)
```
this is suspicious. some sort of comment for next maintainer, debugger about all these nils?

aside for later - should be using options or config struct to make the intention clearer?

cedric-cordenier replied:
Done :) Yes, ideally we'd put these handler deps in a separate config struct

krehermann commented on core/services/gateway/handlers/vault/aggregator.go:
```diff
@@ -0,0 +1,148 @@
+		return currResp, nil
+	}
+
+	l.Debugw("failed to validate signatures, falling back to quorum aggregation", "error", err)
```
add a field to the logger for the request id or similiar for monitoring and debugging

cedric-cordenier replied:
This is already happening -- I import the logger from the caller that includes the RequestID, method and some other debug fields.

krehermann commented on core/services/gateway/handlers/vault/aggregator.go:
```diff
@@ -0,0 +1,148 @@
+		// TODO: not sure this digest is correct -- we need to exclude signatures if they exist.
+		sha, err := a.sha(r)
+		if err != nil {
+			return nil, fmt.Errorf("failed to validate using quorum: failed to compute digest: %w", err)
```
should this really return? can't you tolerate some number of such errors (maybe f?)?

krehermann replied:
maybe we need a logger and warn or error, including the request id

cedric-cordenier replied:
Done!

krehermann commented on core/services/gateway/handlers/vault/handler.go:
```diff
@@ -257,10 +274,24 @@ func (h *handler) HandleNodeMessage(ctx context.Context, resp *jsonrpc.Response[
+
+	resp, err := h.aggregator.Aggregate(ctx, l, ar, resp)
+	if err != nil {
+		l.Debugw("error aggregating responses", "error", err)
```
warn or error level instead?

krehermann commented on core/services/registrysyncer/local_registry.go:
```diff
@@ -171,6 +171,59 @@ func (l *LocalRegistry) NodeByPeerID(ctx context.Context, peerID types.PeerID) (
+			if cid == capabilityID {
+				nodes, err := l.nodesForDON(ctx, don.DON)
+				if err != nil {
+					return nil, err
```
wrap err with capability id for debugging

cedric-cordenier committed:
Address feedback

prashantkumar1982 commented on core/services/gateway/handlers/vault/aggregator.go:
```diff
@@ -0,0 +1,148 @@
+	}
+
+	// Case: We have signatures. In this case we copy the response,
+	// zeroing out the signatures, and take the resulting digest.
```
sha() is only used for quorum.
So isn't it better to just take the payload for sha, and ignore both signatures and context?

cedric-cordenier replied:
Hmmmm 🤔 I would rather leave it in since we would expect the context to match all requests. The context includes OCR epoch + round information so it guarantees that all responses will have come from the same round.

Otherwise you could imagine a scenario where a malicious node responds with a report across two separate rounds and can therefore lead to inclusion

Thinking about it more... we probably need to ensure that we only allow 1 response per node.

prashantkumar1982 commented on core/services/gateway/handlers/vault/handler.go:
```diff
@@ -275,34 +300,45 @@ func (h *handler) HandleNodeMessage(ctx context.Context, resp *jsonrpc.Response[
 	}
-	if createSecretsRequests.RequestId == "" {
+
+	createSecretsRequest.RequestId = ar.req.ID
```
I'm beginning to believe we shouldn't have 2 places to put reqId for the user.
Here, if user puts different ids on json message vs here, we are silently ignoring this inner id without validating.

Instead we should harden our spec so that id can be provided by user only in 1 place, so no place for any confusion.

Just wondering out loud for later, nothing to change in this PR.

cedric-cordenier replied:
I agree 👍 I think the answer could be to split out the user-facing spec from the internal spec, with the difference being that the user-facing spec is the same as the proto message but without the requestID (since the ID will be provided by the user in the standard jsonrpc field)

Wdyt?

prashantkumar1982 replied:
Yes. I would also want to remove the Owner field from user facing spec.
Because owner is generated during our auth process, and there's no point letting the user provide it. We will anyways not trust the owner fied that user provides.

prashantkumar1982 commented on core/services/ocr2/delegate.go:
```diff
@@ -706,7 +705,7 @@ func (d *Delegate) newServicesVaultPlugin(
 	srvs = append(srvs, handler)
 
-	if gwerr := gwconnector.AddHandler(ctx, []string{vaultapi.MethodSecretsCreate, vaultapi.MethodSecretsGet, vaultapi.MethodSecretsUpdate, vaultapi.MethodSecretsDelete, vaultapi.MethodSecretsList}, handler); gwerr != nil {
+	if gwerr := gwconnector.AddHandler(ctx, vaultcap.Methods, handler); gwerr != nil {
```
nice!
However a note, we need to block Get method from gateway. It was only added for testing.

cedric-cordenier replied:
Good point 👍 I'll remove GetSecrets from this list and add a comment saying it can be added for testing purposes

prashantkumar1982 commented on core/services/gateway/handlers/vault/aggregator.go:
```diff
@@ -0,0 +1,148 @@
+			return r, nil
+		}
+	}
+
```
We likely need to fail faster here and handle that case.
If maximum shaToCount value + remaining responses is still less than requiredQuorum, we need to give an error indicating this request fails quorum.

As I believe, right now, if we don't get a quorum even after all responses are available, we don't respond back to user right?
Does the user see a timeout currently?

prashantkumar1982 commented on core/services/gateway/handlers/vault/handler.go:
```diff
@@ -257,10 +268,24 @@ func (h *handler) HandleNodeMessage(ctx context.Context, resp *jsonrpc.Response[
+	ar.mu.Unlock()
+
+	resp, err := h.aggregator.Aggregate(ctx, l, ar, resp)
+	if err != nil {
```
We need to categorize errors:
1. Not enough responses yet to achieve quorum, so still waiting to respond to user
2. Eventual quorum not possible. Send this response to user

prashantkumar1982 commented on core/services/gateway/handlers/vault/aggregator.go:
```diff
@@ -0,0 +1,148 @@
+	capabilitiesRegistry capabilitiesRegistry
+}
+
+func (a *baseAggregator) Aggregate(ctx context.Context, l logger.Logger, ar *activeRequest, currResp *jsonrpc.Response[json.RawMessage]) (*jsonrpc.Response[json.RawMessage], error) {
```
We need another condition where if total responses are equal to total don members, then we respond to user immediately, even if we don't have a quorum based or signature based response.

Basically to handle bad response case more elegantly for user.

I suspect currently in these cases, we are letting the request time out.

cedric-cordenier replied:
Correct -- we are

Let me rework this a bit

prashantkumar1982 reviewed (commented):
Very nicely done!
Few comments on fast failing around aggregation logic, but otherwise looks very neat!!

cedric-cordenier committed:
Address feedback

- Comment out GetSecrets from the method list
- Handle some aggregation edge cases

cedric-cordenier committed:
Merge branch 'develop' into PRIV-128-aggregate-responses

cedric-cordenier committed:
Remove deleted file

prashantkumar1982 commented on core/services/gateway/handlers/vault/handler.go:
```diff
@@ -269,12 +271,22 @@ func (h *handler) HandleNodeMessage(ctx context.Context, resp *jsonrpc.Response[
-		l.Warnw("error aggregating responses", "error", err)
+	switch {
+	case errors.Is(err, errQuorumUnobtainable):
+		l.Error("quorum unobtainable, returning response to user...", "error", err)
```
Should this be a critical error?
It almost certainly requires manual intervention from CRE engineers to see why quorum isn't reachable, and liekly almost always a core bug in our implementation?

Also, we should log all responses verbose too with this log, to ease troubleshooting. Cause the logical step would be to check responses and see why they aren't all the same.

prashantkumar1982 commented on core/services/gateway/handlers/vault/handler.go:
```diff
@@ -269,12 +271,22 @@ func (h *handler) HandleNodeMessage(ctx context.Context, resp *jsonrpc.Response[
+		l.Error("quorum unobtainable, returning response to user...", "error", err)
+		return h.sendResponse(ctx, ar, h.errorResponse(ar.req, api.FatalError, err))
+	case err != nil:
+		l.Warnw("error aggregating responses, waiting for other nodes...", "error", err)
```
This should be debug or info log, coz it's expected to be logged for each request multiple times before reaching quorum.
